Readme
Flamez IRC Services -- a system of IRC services for IRC networks
---------------------------------------------------------

There is absolutely NO WARRANTY provided with this program; 
if it blows up in your face, you get to clean up the mess.  
Services may be freely redistributed; see the GNU 
General Public License (in the file "COPYING") for details.


TABLE OF CONTENTS
      1. Credits
      2. Introduction
      3. Configuration, Compilation and Installation
      4. Operation
      5. Overview of Services Clients
      6. IRC Protocol Additions
      7. Importing Databases From Other Programs
      8. Reaching The Maintainer
      9. Internal Security

The latest version of Services can be obtained from the places below.
The Services web site at:
        http://www.flamez.net/

1. CREDITS

     Flamez-IRCServices developed by and copyright (c) 2001-2002

     We would like to thank all the users who have stuck
by us and continue to help us with improving these
services. Many thanks to all who have helped make
another services release a possibility. Please visit
us on http://www.flamez.net and report any bugs or new
ideas to support@flamez.net
	
     Author: jabea - jabea@insiderz.net

     Project Leader: gcc - gcc@insiderz.net

     NOTE: These services are a modification of ircservices
by A. Church. Special thanks go to him for his inspiration
to open source his services. (http://www.ircservices.za.net/)


2. INTRODUCTION

     IRC Services is a system of services (as the name implies) to be used
with Internet Relay Chat networks.  Services provides for definitive
nickname and channel ownership, as well as the ability to send messages
("memos") to offline users, and gives IRC operators considerably more
control over the network, enabling them to change modes in any channel and
place network-wide bans, among other things.

     Services runs as a server on an IRC network, and is designed to use
the RFC 1459 IRC protocol, with some optional additions discussed at the
end of this document.

     Services was originally designed for use with versions of the DALnet
IRC server implementation (ircd.dal) through 4.4.13.  Currently, Services
interoperates with the following IRC servers:
          ircd-2.8.x (or any RFC1459-compliant server)
          ircd-2.8.x+TS8
          ircu 2.9.x
          ircd.dal 4.4.x
          ircd.dal 4.6.x (Dreamforge)
          Bahamut 1.4.23 and later
          Unreal 3.1.1 (DEVELOPMENT, UNSTABLE)

Services has also been reported to work with UltimateIRCD 2.8.1, using IRC
server type 22 (Dreamforge) in the configure script, and with Chunky Monkey
IRCD 0.9 or later, using IRC server type 23 (Bahamut).

The following servers have been reported NOT to work with Services:
          NewNet
          ircd-hybrid
          ircd 2.x with "+CS" extension
          TS4
          ircd 2.9.x
          ircd 2.10.x

Services also does not support the IRC protocol extensions defined in RFCs
2811-2813.

     If you have one of these servers or you cannot get Services to work
with your server, please switch to one of the supported servers listed
above.


3. CONFIGURATION, COMPILATION AND INSTALLATION

     In order to compile Services, you'll need the Bourne shell or a
compatible shell (such as GNU bash), GNU make or a compatible make (which
needs to support the "include" and "ifdef" directives), and an ANSI C
compiler (gcc 2.x recommended; also see (*) below).  If you want to modify
the language files, you will also need the Perl interpreter in your path.
All GNU utilities can be found at ftp://prep.ai.mit.edu/pub/gnu/.
NOTE: The "make" distributed with FreeBSD is known not to work with
Services; use GNU make instead, usually installed as "gmake".

(*) Services is coded (loosely) around the ANSI standard, and will not
    compile under K&R compilers.  However, Services does not attempt
    complete ANSI conformance, and may fail to compile or run correctly
    using some compilers that claim to be ANSI-compatible.  In particular,
    gcc 3.0 reportedly plays games with structure member ordering (this
    actually appears to violate ANSI) and should be avoided for the time
    being.

     Before trying to compile Services, there are some manual configuration
tasks that need to be done: run the "configure" script, (optionally) edit
config.h, and (optionally) edit the top section of the Makefile.

     The "configure" script will try to learn how things work on your
system, and set up appropriate Makefile and C source definitions.  It will
also ask you a few questions about how you want Services set up.  (It will
ask all of the questions before doing any of the automated tests.)  If you
get an error while running the script, get bash (if you don't already have
it) and run "bash configure".  IMPORTANT NOTE:  Make sure you select the
correct IRC server style!  If Services compiles correctly but you only get
messages like "Internal error - unable to process request" or "ERROR Wrong
protocol", the most likely cause is that you chose the wrong IRC server
style.

     Note that when asked for the binary and data file installation
directories, you need to choose directories other than the Services source
directory (if you try to install to the Services source directory, the
configure script will display an error message).  If you are installing
Services as a normal user (as opposed to root) and do not have write access
to the default /usr/local/sbin and /usr/local/lib directories, I would
suggest using "bin" and "lib" directories under your home directory, or
creating a separate Services directory and installing the files there.
You may choose the same directory for both settings if you so desire.

     If you are compiling Services for use on a different system, or in
another situation where you cannot or do not want to create the directories
given for the install script, run the configure script with the
"-no-dir-check" option.  This will omit the check for existence of the
specified directories.

     config.h contains a few basic Services settings; in most cases you
should not need to change these.  Most settings will be configured via the
"services.conf" file (discussed below).

     The Makefile has a section at the top allowing you to configure
certain compile-time options.  You can also add additional C compiler flags
here, such as warning or optimization flags (this can also be done in the
configure script using the -cflags command-line option).

     Once these steps are done, you should be able to compile Services with
little trouble.  If Services fails to compile on your system, or if you
have to tweak configure's output to get it to work, please send a bug
report so that the problem can be fixed for future releases.

     Once Services is compiled, type "make install" to copy the program and
data files to their destinations.  Care should be taken that only
authorized people have access to the data files; by default (changeable
via the configure script), passwords are NOT encrypted, so unauthorized
access to the files would be a big problem!

     Finally, if you are using Services for the first time or upgrading
from a version prior to 4.1.0, you will need to create the Services
configuration file.  An example configuration file is provided in the file
"example.conf", which will be installed in the Services data directory.
Edit this file to your liking (note that you will need to fill in values
for at least some directives--pay particular attention to ServicesRoot),
then rename the file to "services.conf".

     If you are upgrading from an earlier version of Services, check the
WhatsNew file for a list of options which have been added.  Descriptions of
the options can be found in the example.conf file.


4. OPERATION

     This section does not detail the operation of the individual pieces of
Services; that information can be found in the online help files
("/msg *Serv help").  It only describes how to start Services itself.

     Normally, Services can be run simply by invoking the "services"
executable.  Services will then use the defaults specified in the
services.conf file, and connect to the specified uplink server.
Alternatively, any of the following command-line options can be specified
to override the default values:

        -remote server[:port]   Connect to the specified server
        -local host  -or-       Connect from the specified address (e.g.
               [host]:[port]        for multihomed servers)
        -name servername        Our server name (e.g. services.some.net)
        -desc string            Description of us (e.g. SomeNet Services)
        -numeric numeric        (Unreal) Server numeric, 1-254 or 0 for none
        -user username          Username for Services' nicks (e.g. services)
        -host hostname          Hostname for Services' nicks (e.g. some.net)
        -dir directory          Directory containing Services' data files
                                    (e.g. /usr/local/lib/services)
        -log filename           Services log filename (e.g. services.log)
        -update secs            How often to update databases (in seconds)
        -expire secs            How often to check for nick/channel
                                    expiration (in seconds)

     Additionally, the following command-line options can be used to modify
the behavior of Services:

        -debug                  Enable debugging mode--more info sent to log
                                    (give option more times for more info)
        -readonly               Enable read-only mode--no changes to
                                    databases allowed, .db files and log
                                    not written
        -skeleton               Enable skeleton mode--like read-only mode,
                                    but only OperServ is available
        -nofork                 Do not fork after startup; log messages will
                                    be written to terminal (as well as to
                                    the log file if not in read-only mode)
        -forceload              Try to load as much of the databases as
                                    possible, even if errors are encountered
        -noexpire               Prevents all expirations (nicknames, channels,
                                    akills, session limit exceptions, etc.)
        -noakill                Disables autokill checking

     The above list of options can be obtained using the "-help" option
(which may be abbreviated to "-h").

     Upon starting, Services will parse its command-line parameters, open
its logfile, then (assuming the -nofork option is not given) detach itself
and run in the background.  If Services encounters a problem reading the
database files or cannot connect to its uplink server, it will terminate
immediately; otherwise, it will run until the connection is terminated (or
a QUIT, SHUTDOWN, or RESTART command is sent--see OperServ's help).  In the
case of an error, an appropriate error message will be written to the log
file.  Services may also be restarted from outside IRC by sending it the
SIGHUP signal.

     If Services is run with the "-readonly" command-line option, it can
serve as a "backup" to the full version of Services.  A "full" version of
Services (run without -readonly) will automatically reintroduce its
pseudo-clients (NickServ, ChanServ, etc.), while a "backup" Services will
not, thus allowing full Services to be brought up at any time without
disrupting the network (and without having to take backup Services down
beforehand).

     If Services is run with the "-skeleton" command-line option, it will
not try to load the nickname or channel databases, and will respond with
"services is inactive" messages to any commands sent to NickServ, ChanServ,
or MemoServ.  This can be useful as an emergency stopgap measure when the
main copy of Services cannot be started.

     The "-debug" option is useful if you find or suspect a problem in
Services. Giving it once on the command line will cause all traffic to and
from Services as well as some other debugging information to be recorded in
the log file; if you send a bug report, PLEASE include an excerpt from the
log file WITH DEBUGGING ACTIVE--I cannot emphasize enough how important
this is to tracking down problems.  (You can also enable debugging while
Services is running using OperServ's SET DEBUG command.)  If you repeat the
-debug option more than once, the debugging level will be increased, which
provides more detailed information but may also slow Services down
considerably and make the log file grow dramatically faster (in particular,
at debug level 4 a message is written to the log for every character
received from the server).  In general, a debug level of 1 is sufficient
to trace a problem, because all network traffic is included and I can
usually reproduce the problem.

     The "-forceload" option is provided to attempt recovery of data from
corrupted or truncated databases.  Normally, if Services encounters an
error writing to a database file, it will attempt to restore the original
version of the file and report an error to the logfile and through WALLOPS.
However, if this should fail (which normally should not happen), or if
Services is terminated abruptly e.g. by kill -9 or a power failure, then
one or more of the databases may be corrupt.  Normally, this will cause
Services to abort the next time you try to run it; however, if you give
the -forceload option to Services, it will instead read as much as it can,
then skip to the next database.  For obvious reasons, it's recommended to
keep backup copies of your databases in case something does happen (since
Services will stop at the first error even with -forceload, meaning you
lose any data after that).

     Two additional programs are available in addition to the main
executable: "listnicks" and "listchans", both installed as hard links to
the main executable.  The programs will list all registered nicknames and
channels, respectively; or, if given the -c option, will display the
number of registered nicknames or channels.


5. OVERVIEW OF SERVICES CLIENTS

     This is a brief introduction to the various clients available from
Services.  All *Serv clients can be /msg'd with "help" for a general
introduction or "help <command>" for more detailed command descriptions.

     NickServ is the nickname server; it allows users to register and
control nicknames, and will (at the user's choice) /kill any unauthorized
user who tries to use that nick after a warning.  NickServ also allows
users to select a language which Services will then use for all
communication with the user (command responses as well as automatic
notices).

     ChanServ is the channel server; as NickServ does with nicknames, it
allows users to register and control channels.  There is a much wider array

Changes
Version 4.5
-----------
2002/06/24 .41	Fixed format-string crash bug reported by Lucas Nussbaum
		    <lucas@lucas-nussbaum.net>
2002/05/10 .40	Fixed bug causing some mode changes to get lost with
		    MergeChannelModes.  Reported by <fabulous@t7ds.com.br>
2002/02/27 .39	Fixed minor security hole allowing users to find out when
		    nicknames not their own receive memos.
2002/02/15	Fixed ChanServ LIST syntax error message for Services
		    admins.  Reported by Mark Hetherington <mark@ctcp.net>
2002/02/07 .38	Fixed bug causing AUTODEOP channel level to be ignored if
		    AUTOVOICE was set less or equal.  Reported by Jollino
		    <jollino@sogno.net>
2002/01/14 .37	Fixed a bug in the Japanese language file causing crashes.
		    Reported by <fabulous@t7ds.com.br>
2002/01/12	Fixed a bug causing the SET HIDE USERMASK setting for
		    nicknames to get lost when importing Sirv or Auspice
		    databases.
2002/01/07 .36	Forbid time is now recorded with forbidden nicknames and
		    channels.
2002/01/03	Removed a debugging message that could sometimes get sent
		    from MemoServ INFO.
2001/12/17 .35	Added missing support for Unreal CHGIDENT/CHGHOST/CHGNAME.
		    Reported by Michael Raff <raff@electrocity.com>
2001/12/17	Fixed bug in TIME reply.  Reported by <v13@priest.com>
2001/12/17	Fixed typo in Japanese language file causing crashes.
		    Reported by <fabulous@t7ds.com.br>
2001/12/03	Fixed bug in import-db preventing compilation when an IRC
		    server type other than Unreal was selected.  Reported
		    by Dung T. Nguyen <theonlydung@rogers.com>
2001/12/02	Fixed bug in autokill checking causing permanent autokills
		    to expire.  Reported by Lloyd Williams
		   <binary_frog@chatcircuit.com>
2001/12/01 .34	Fixed cosmetic bug with autokill expiration wallops.
		    Reported by Lloyd Williams <binary_frog@chatcircuit.com>
2001/12/01	Fixed typo in Turkish language file.  Reported by Guven
		    Guzelbey <guzelbey@cs.utk.edu>
2001/11/28	Fixed bug in Unreal TKL handling causing permanent autokills
		    to expire.  Reported by Tom Moyer <squall157@hotmail.com>
2001/11/26	Fixed bug in importing Sirv databases.  Reported by Hakan
		    Karakan <hkarakan@zurna.net>
2001/11/23 .33	Updated E-mail addresses in documentation.
2001/11/17	Autokicks now remove ban exceptions from channels when
		    triggered.  Suggested by <RealCFC@chatfirst.com>
2001/11/17	Corrected typo in OperServ KILLCLONES help.  Reported by
		    Ali Sor <alisor@softhome.net>
2001/11/16 .32	433 is no longer sent for users who have identified.
		    Reported by Michael Sparpaglione <cyxs@cyxs.net>
2001/11/12 .31	Changed 433 handling to delay a bit before sending.
2001/11/12	Fixed bug causing crashes on FORBID of logged-on nick.
		    Reported by <fabulous@t7ds.com.br>
2001/11/07	Added support for +M channel mode in Bahamut 1.4.30.
		    Suggested by Andrew Kempe <andrewk@isdial.net>
2001/11/06 .30	Fixed bug which left expired autokills in force.
2001/11/06	Added handling for Unreal TKL messages to avoid expired or
		    deleted autokills reappearing in the network.  Reported
		    by <ron885@linuxfreemail.com>
2001/11/06	Added sending of 433 (nick in use) on regged nick warnings.
		    Suggested by Finny Merrill <griever@ircd-net.org>
2001/11/01	Fixed a bug causing list{nicks,chans} -d <dir> to not work.
2001/10/31	Fixed incorrect check for nick suspension in NickServ
		    IDENTIFY.  Reported by Lonewolf <lonewolf@lagnet.org.za>
2001/10/29	Fixed incorrect channel count check in NickServ LINK.
2001/10/28 .29	Fixed autokills to no longer kill users after the
		    expiration time but before actually being expired.
2001/10/28	Added support for the Unreal TKL (timed K:line) message.
2001/10/27	Added workaround to use DEF_LANGUAGE for forbidden nick
		    messages.  Reported by Ali Sor <alisor@softhome.net>
2001/10/27	Fixed bug causing crashes when a user's server is unknown
		    to Services (e.g. because hostmasking is in use).
		    Reported by GB_RAIDER1 <raider1@clan-ghost-bear.com>
2001/10/01 .28	Added Dutch language file, courtesy of Martin Pels
		    <rodecker@mp3crew.nu>
2001/09/14 .27	Fixed bug in configure script causing compilation errors.
		    Reported by Scott Grayban <borg@borgdrone.net>
2001/09/14 .26	Fixed bug causing channel mode locks to be lost when
		    importing Magick 1.4 or SirvNET Services databases.
2001/09/14	Added support for Auspice 2.5 databases to import-db.
		    Suggested by MX <darkie@tampabay.rr.com>
2001/09/11	INFO was not listed in the MemoServ help message; added it.
		    Reported by Martin Pels <martinpels@hotmail.com>
2001/09/11	Fixed fakehost-related (Unreal) bug sometimes triggered by
		    ChanServ bans.  Reported by Carsten V. Munk
		    <stskeeps@tspre.org>
2001/09/04 .25	Fixed bug in Turkish language file causing crashes.
2001/08/14	Added missing LISTCHANS to NickServ help.  Reported by
		    <gousoufou@hushmail.com>
2001/08/13	Fixed bug allowing unsetting of E-mail address even when
		    NSRequireEmail is set.
2001/08/12 .24	Fixed bug preventing deletion of no-longer-existent servers
		    from StatServ.  Reported by Josh
		    <greenday288@geocities.com>
2001/08/10	Corrected a security-related comment in nickserv.c.
2001/08/09	Added missing support for PTlink autokill databases to
		    import-db.  Reported by Yusuf Iskenderoglu
		    <uhc0@stud.uni-karlsruhe.de>
2001/08/07 .23	Fixed bug allowing evasion of nick kill on linked nicks.
		    Reported by Georges Berscheid <Georges@berscheid.lu>
2001/08/02 .22	Expanded import-db Epona support to support versions 1.3.0
		    through 1.4.0.
2001/08/02	Expanded import-db SirvNET support to support all versions.
2001/07/31	Fixed bug converting news database from Daylight.
2001/07/30	Added support for Wrecked IRC Services 1.2.0 databases to
		    import-db.  Suggested by Jamie Smithson
		    <jamie@silverdream.org.uk>
2001/07/29	Fixed bug in import-db causing conversion to fail if there
		    were no .db files already in place.

init.c
/* Initalization and related routines.
 *
 * IRC Services is copyright (c) 1996-2002 Andrew Church.
 *     E-mail: <achurch@achurch.org>
 * Parts copyright (c) 1999-2000 Andrew Kempe and others.
 * This program is free but copyrighted software; see the file COPYING for
 * details.
 */

#include "services.h"

/*************************************************************************/

/* Send a NICK command for the given pseudo-client.  If `user' is NULL,
 * send NICK commands for all the pseudo-clients. */

#ifdef IRC_UNREAL
# define EXTRA_MODES (UMODE_S | UMODE_q | UMODE_d)
#else
# define EXTRA_MODES 0
#endif
#define NICK(nick,name,modes) \
    send_nick((nick), ServiceUser, ServiceHost, ServerName, (name), (modes) | EXTRA_MODES);

void introduce_user(const char *user)
{
    /* Watch out for infinite loops... */
#define LTSIZE 20
    static int lasttimes[LTSIZE];
    if (lasttimes[0] >= time(NULL)-3)
	fatal("introduce_user() loop detected");
    memmove(lasttimes, lasttimes+1, sizeof(lasttimes)-sizeof(int));
    lasttimes[LTSIZE-1] = time(NULL);
#undef LTSIZE

    if (!user || stricmp(user, s_NickServ) == 0)
	NICK(s_NickServ, desc_NickServ, UMODE_o);
    if (!user || stricmp(user, s_ChanServ) == 0)
	NICK(s_ChanServ, desc_ChanServ, UMODE_o);
    if (!user || stricmp(user, s_HelpServ) == 0)
	NICK(s_HelpServ, desc_HelpServ, 0);
    if (s_IrcIIHelp && (!user || stricmp(user, s_IrcIIHelp) == 0))
	NICK(s_IrcIIHelp, desc_IrcIIHelp, 0);
    if (!user || stricmp(user, s_MemoServ) == 0)
	NICK(s_MemoServ, desc_MemoServ, UMODE_o);
    if (!user || stricmp(user, s_OperServ) == 0)
	NICK(s_OperServ, desc_OperServ, UMODE_o | UMODE_i);
#ifdef STATISTICS
    if (!user || stricmp(user, s_StatServ) == 0)
	NICK(s_StatServ, desc_StatServ, UMODE_i);
#endif
    if (s_DevNull && (!user || stricmp(user, s_DevNull) == 0))
	NICK(s_DevNull, desc_DevNull, UMODE_i);
    if (!user || stricmp(user, s_GlobalNoticer) == 0)
	NICK(s_GlobalNoticer, desc_GlobalNoticer, UMODE_o | UMODE_i);
	if (!user || stricmp(user, s_FloodServ) == 0)
	NICK(s_FloodServ, desc_FloodServ, UMODE_i);
}

#undef NICK

/*************************************************************************/

/* Set GID if necessary.  Return 0 if successful (or if RUNGROUP not
 * defined), else print an error message to logfile and return -1.
 */

static int set_group(void)
{
#if defined(RUNGROUP) && defined(HAVE_SETGRENT)
    struct group *gr;

    setgrent();
    while ((gr = getgrent()) != NULL) {
	if (strcmp(gr->gr_name, RUNGROUP) == 0)
	    break;
    }
    endgrent();
    if (gr) {
	setgid(gr->gr_gid);
	return 0;
    } else {
	log("Unknown group `%s'\n", RUNGROUP);
	return -1;
    }
#else
    return 0;
#endif
}

/*************************************************************************/

/* Parse command-line options.  Return 0 if all went well, -1 for an error
 * with an option, or 1 for -help.
 */

static int parse_options(int ac, char **av)
{
    int i, n;
    char *s, *t;

    debug = 0;
    for (i = 1; i < ac; i++) {
	s = av[i];
	if (*s == '-') {
	    s++;
	    if (strcmp(s, "remote") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-remote requires hostname[:port]\n");
		    return -1;
		}
		s = av[i];
		t = strchr(s, ':');
		if (t) {
		    *t++ = 0;
		    if (atoi(t) > 0)
			RemotePort = atoi(t);
		    else {
			fprintf(stderr, "-remote: port number must be a positive integer.  Using default.");
			return -1;
		    }
		    t[-1] = ':';
		}
		RemoteServer = s;
	    } else if (strcmp(s, "local") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-local requires hostname or [hostname]:[port]\n");
		    return -1;
		}
		s = av[i];
		t = strchr(s, ':');
		if (t) {
		    *t = 0;
		    n = atoi(t+1);
		    if (n >= 0)
			LocalPort = n;
		    else
			fprintf(stderr, "-local: port number must be a positive integer or 0.  Using default.\n");
		    *t = ':';
		}
		LocalHost = s;
	    } else if (strcmp(s, "name") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-name requires a parameter\n");
		    return -1;
		}
		ServerName = av[i];
	    } else if (strcmp(s, "desc") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-desc requires a parameter\n");
		    return -1;
		}
		ServerDesc = av[i];
	    } else if (strcmp(s, "numeric") == 0) {
#ifdef IRC_UNREAL
		if (++i >= ac) {
		    fprintf(stderr, "-numeric requires a parameter\n");
		    return -1;
		}
		n = atoi(av[i]);
		if (n < 0 || n > 254)
		    fprintf(stderr, "-numeric parameter must be between 0 and 254 inclusive.  Using default.\n");
		else
		    ServerNumeric = n;
#else
		fprintf(stderr, "-numeric available only with Unreal IRC daemon\n");
		return -1;
#endif
	    } else if (strcmp(s, "user") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-user requires a parameter\n");
		    return -1;
		}
		ServiceUser = av[i];
	    } else if (strcmp(s, "host") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-host requires a parameter\n");
		    return -1;
		}
		ServiceHost = av[i];
	    } else if (strcmp(s, "dir") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-dir requires a parameter\n");
		    return -1;
		}
		services_dir = av[i];
	    } else if (strcmp(s, "log") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-log requires a parameter\n");
		    return -1;
		}
		log_filename = av[i];
	    } else if (strcmp(s, "update") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-update requires a parameter\n");
		    return -1;
		}
		s = av[i];
		if (atoi(s) <= 0) {
		    fprintf(stderr, "-update: number of seconds must be positive");
		    return -1;
		} else
		    UpdateTimeout = atol(s);
	    } else if (strcmp(s, "expire") == 0) {
		if (++i >= ac) {
		    fprintf(stderr, "-expire requires a parameter\n");
		    return -1;
		}
		s = av[i];
		if (atoi(s) <= 0) {
		    fprintf(stderr, "-expire: number of seconds must be positive");
		    return -1;
		} else
		    ExpireTimeout = atol(s);
	    } else if (strcmp(s, "debug") == 0) {
		debug++;
	    } else if (strcmp(s, "readonly") == 0) {
		readonly = 1;
		skeleton = 0;
	    } else if (strcmp(s, "skeleton") == 0) {
		readonly = 0;
		skeleton = 1;
	    } else if (strcmp(s, "nofork") == 0) {
		nofork = 1;
	    } else if (strcmp(s, "forceload") == 0) {
		forceload = 1;
	    } else if (strcmp(s, "noexpire") == 0) {
		noexpire = 1;
	    } else if (strcmp(s, "noakill") == 0) {
		noakill = 1;
	    } else if (strcmp(s, "h") == 0 || strcmp(s, "help") == 0
		       || strcmp(s, "-help") == 0) {
		fputs("\
The following options are recognized:
	-remote server[:port]	Connect to the specified server
	-local host[:port]	Connect from the specified address (e.g.
				    for multihomed servers)
	-name servername	Our server name (e.g. services.some.net)
	-desc string		Description of us (e.g. SomeNet Services)\n"
#ifdef IRC_UNREAL
"	-numeric numeric	Services server numeric (1-254 or 0 for none)\n"
#endif
"	-user username		Username for Services' nicks (e.g. services)
	-host hostname		Hostname for Services' nicks (e.g. some.net)
	-dir directory		Directory containing Services' data files
				    (e.g. /usr/local/lib/services)
	-log filename		Services log filename (e.g. services.log)
	-update secs		How often to update databases (in seconds)
	-expire secs		How often to check for nick/channel
				    expiration (in seconds)
	-debug			Enable debugging mode--more info sent to log
				    (give option more times for more info)
	-readonly		Enable read-only mode--no changes to
				    databases allowed, .db files and log
				    not written
	-skeleton		Enable skeleton mode--like read-only mode,
				    but only OperServ is available
	-nofork			Do not fork after startup; log messages will
				    be written to terminal (as well as to
				    the log file if not in read-only mode)
	-forceload		Try to load as much of the databases as
				    possible, even if errors are encountered
	-noexpire		Prevents all expirations (nicknames, channels,
				    akills, session limit exceptions, etc.)
	-noakill		Disables autokill checking
", stdout);
		exit(0);
	    } else {
		fprintf(stderr, "Unknown option -%s.  Use \"-help\" for help.\n", s);
		return -1;
	    }
	} else {
	    fprintf(stderr, "Non-option arguments not allowed\n");
	    return -1;
	}
    }
    return 0;
}

/*************************************************************************/

/* Remove our PID file.  Done at exit. */

static void remove_pidfile(void)
{
    remove(PIDFilename);
}

/*************************************************************************/

/* Create our PID file and write the PID to it. */

static void write_pidfile(void)
{
    FILE *pidfile;

    pidfile = fopen(PIDFilename, "w");
    if (pidfile) {
	fprintf(pidfile, "%d\n", (int)getpid());
	fclose(pidfile);
	atexit(remove_pidfile);
    } else {
	log_perror("Warning: cannot write to PID file %s", PIDFilename);
    }
}

/*************************************************************************/

/* Overall initialization routine.  Returns 0 on success, -1 on failure. */

int init(int ac, char **av)
{
    int i;
    int openlog_failed = 0, openlog_errno = 0;
    int started_from_term = isatty(0) && isatty(1) && isatty(2);


    /* Initialize pseudo-random number generator. */
    srand(time(NULL) ^ getppid() ^ getpid()<<16);

    /* Set file creation mask and group ID. */
#if defined(DEFUMASK) && HAVE_UMASK
    umask(DEFUMASK);
#endif
    if (set_group() < 0)
	return -1;

    /* Parse command-line options; exit if an error occurs. */
    if (parse_options(ac, av) < 0)
	return -1;

    /* Chdir to Services data directory. */
    if (chdir(services_dir) < 0) {
	fprintf(stderr, "chdir(%s): %s\n", services_dir, strerror(errno));
	return -1;
    }

    /* Open logfile, and complain if we didn't. */
    if (open_log() < 0) {
	openlog_errno = errno;
	if (started_from_term) {
	    fprintf(stderr, "Warning: unable to open log file %s: %s\n",
			log_filename, strerror(errno));
	} else {
	    openlog_failed = 1;
	}
    }

    /* Read configuration file; exit if there are problems. */
    if (!read_config())
	return -1;

    /* Re-parse command-line options (to override configuration file). */
    parse_options(ac, av);

    /* Detach ourselves if requested. */
    if (!nofork) {
	if ((i = fork()) < 0) {
	    perror("fork()");
	    return -1;
	} else if (i != 0) {
	    exit(0);
	}
	if (started_from_term) {
	    close(0);
	    close(1);
	    close(2);
	}
	if (setpgid(0, 0) < 0) {
	    perror("setpgid()");
	    return -1;
	}
    }

#ifdef MEMCHECKS
    /* Account for runtime memory.  Do this after forking to avoid a bogus
     * "XXX bytes leaked on exit" message when the parent exits. */
    init_memory();
#endif

    /* Write our PID to the PID file. */
    write_pidfile();

    /* Announce ourselves to the logfile. */
    if (debug || readonly || skeleton || noexpire) {
	log("Services %s (compiled for %s) starting up (options:%s%s%s%s)",
	    version_number, version_protocol,
	    debug ? " debug" : "",
	    readonly ? " readonly" : "",
	    skeleton ? " skeleton" : "",
	    noexpire ? " noexpire" : "");
    } else {
	log("Services %s (compiled for %s) starting up",
	    version_number, version_protocol);
    }
    start_time = time(NULL);

    /* If in read-only mode, close the logfile again. */
    if (readonly)
	close_log();

    /* Set signal handlers.  Catch certain signals to let us do things or
     * panic as necessary, and ignore all others.
     */
#ifdef NSIG
    for (i = 1; i <= NSIG; i++) {
#else
    for (i = 1; i <= 32; i++) {
#endif
	if (i != SIGPROF)
	    signal(i, SIG_IGN);
    }

    signal(SIGINT, weirdsig_handler);
    signal(SIGTERM, weirdsig_handler);
    signal(SIGQUIT, weirdsig_handler);
#ifndef DUMPCORE
    signal(SIGSEGV, weirdsig_handler);
#endif
    signal(SIGBUS, weirdsig_handler);
    signal(SIGQUIT, weirdsig_handler);
    signal(SIGHUP, weirdsig_handler);
    signal(SIGILL, weirdsig_handler);
    signal(SIGTRAP, weirdsig_handler);
    signal(SIGFPE, weirdsig_handler);
#ifdef SIGIOT
    signal(SIGIOT, weirdsig_handler);
#endif

    /* This is our "out-of-memory" panic switch */
    signal(SIGUSR1, weirdsig_handler);

    /* Initialize multi-language support */
    lang_init();
    if (debug)
	log("debug: Loaded languages");

    /* Initialiize subservices */
    ns_init();
    cs_init();
    ms_init();
    os_init();

    /* Load up databases */
    if (!skeleton) {
	load_ns_dbase();
	if (debug)
	    log("debug: Loaded %s database (1/13)", s_NickServ);
	load_cs_dbase();
	if (debug)
	    log("debug: Loaded %s database (2/13)", s_ChanServ);
    }
    load_os_dbase();
    if (debug)
	log("debug: Loaded %s database (3/13)", s_OperServ);
    load_akill();
    if (debug)
	log("debug: Loaded AKILL database (4/13)");
    load_news();
    if (debug)
	log("debug: Loaded news database (5/13)");
#ifndef STREAMLINED
    load_exceptions();
    if (debug)
        log("debug: Loaded exception database (6/13)");
#else
    if (debug)
        log("debug: Not loading exception database (6/13)");
#endif
#ifdef STATISTICS
    load_ss_dbase();
    if (debug)
	log("debug: Loaded statistics database (7/13)");
#else
    if (debug)
	log("debug: Not loading statistics database (7/13)");
#endif
    log("Databases loaded");

	load_nooper();
	if (debug)
	log("debug: Loaded NOOPER database (8/13)");
    log("Databases loaded");
	
	load_snooper();
	if (debug)
	log("debug: Loaded SNOOPER database (9/13)");
    log("Databases loaded");
	
	load_aconnect();
	if (debug)
	log("debug: Loaded ACONNECT database (10/13)");
    log("Databases loaded");
    
	load_nakill();
	if (debug)
	log("debug: Loaded NAKILL database (11/13)");
    log("Databases loaded");

	load_fs_dbase();
	if (debug)
	log("debug: Loaded FLOODSERV database (12/13)");
    log("Databases loaded");
    
	load_grname_dbase();
	if (debug)
	log("debug: Loaded GRNAME database (13/13)");
    log("Databases loaded");
 
    /* Connect to the remote server */
    servsock = conn(RemoteServer, RemotePort, LocalHost, LocalPort);
    if (servsock < 0)
	fatal_perror("Can't connect to server");
    send_server();
    sgets2(inbuf, sizeof(inbuf), servsock);
    if (strnicmp(inbuf, "ERROR", 5) == 0) {
	/* Close server socket first to stop wallops, since the other
	 * server doesn't want to listen to us anyway */
	disconn(servsock);
	servsock = -1;
	fatal("Remote server returned: %s", inbuf);
    }

    /* Announce a logfile error if there was one */
    if (openlog_failed) {
	wallops(NULL, "Warning: couldn't open logfile: %s",
		strerror(openlog_errno));
    }

    /* Bring in our pseudo-clients */
    introduce_user(NULL);

	/* Rejoin FS channel, if some are in the DB */
	fs_init();

    /* Success! */
    return 0;
}

/*************************************************************************/
